File: DynamicProgramming/2D_DP/GridDP/StringDP/EditDistance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = [[0 for i in range(len(word1) + 1)] for i in range(len(word2) + 1)]
        for r in range(len(word2) + 1):
            dp[r][0] = r
        for c in range(len(word1) + 1):
            dp[0][c] = c

        for i in range(1, len(word2) + 1):
            for j in range(1, len(word1) + 1):
                if word2[i - 1] == word1[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
        return dp[len(word2)][len(word1)]

File: DynamicProgramming/2D_DP/StringDP/EditDistance.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(word2.size()+1, vector<int>(word1.size()+1, 0));
        for (int r = 0 ; r < word2.size()+1; r++) {
            dp[r][0] = r;
        }
        for (int c = 0; c < word1.size()+1; c++) {
            dp[0][c] = c;
        }
        for (int i = 1; i < word2.size()+1; i++) {
            for (int j = 1; j < word1.size()+1; j++) {
                if (word2[i-1] == word1[j-1]) {
                    dp[i][j] = dp[i-1][j-1];
                }
                else {
                    dp[i][j] = min({dp[i-1][j], dp[i][j-1], dp[i-1][j-1]}) + 1;
                }
            }
        }
        return dp[word2.size()][word1.size()];
    }
};